_posts/2020-01-27-serverless-express-openapi.md

```diff
@@ -0,0 +1,522 @@
+
+We are pretty much done now. Here is the document in all it's glory.
+
+```yaml
+openapi: 3.0.0
+info:
+  title: simple-node-api
+  description: A simple API for a guestbook application
+  version: 2019-10-13
+paths:
+  /:
+    post:
+      operationId: CreateMessage
+      requestBody:
+        content:
+          text/plain:
+            schema:
+              type: string
+        required: true
+      responses:
+        "201":
+          description: "Successfully created a message."
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Message"
+        "400":
+          description: "Bad Request Exception"
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/BadRequestException"
+        "500":
+          description: "Internal Server Error"
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/InternalServerErrorException"
+    get:
+      operationId: ListMessages
+      parameters:
+        - name: maxItems
+          in: query
+          required: false
+          schema:
+            type: number
+        - name: token
+          in: query
+          required: false
+          schema:
+            type: string
+      responses:
+        "200":
+          description: "Successfully listed messages."
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/MessageList"
+        "400":
+          description: "Bad Request Exception"
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/BadRequestException"
+        "500":
+          description: "Internal Server Error"
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/InternalServerErrorException"
+components:
+  schemas:
+    Message:
+      type: object
+      properties:
+        message:
+          type: string
+        author:
+          type: string
+        date:
+          type: string
+          format: date-time
+    MessageList:
+      type: object
+      properties:
+        items: 
+          type: array
+          items:
+            $ref: "#/components/schemas/Message"
+        nextToken:
+          type: string
+          required:
+            - items
+    BadRequestException:
+      type: object
+      properties:
+        message:
+          type: string
+        errorCode:
+          type: string
+    InternalServerErrorException:
+      type: object
+      properties:
+        message:
+          type: string
+        errorCode:
+          type: string
+```
+
+# Adding the Document to SAM
+
+The next thing we need to do is add this document to our SAM template. Here is what the final template will look like (I'll explain the changes after the template).
```

Comment:
You haven't told the reader why we are using sam. Just that "here's a template"... For example:

# Getting our API online using SAM
Now that our API is defined, we need to get it into the AWS cloud. To do that we can use the [AWS SAM](https://aws.amazon.com/serverless/sam/) product which will help bootstrap the infrastructure for us. Let's take a look at the template will look like when we're done. (I explain the changes after the template).